### PR TITLE
[WIP] TensorFlow tutorial notebook

### PR DESCRIPTION
**Pull request by vdumoulin**
@lamblin this is the PR in which I will push updates for the TensorFlow tutorial notebook.

### CODE REVIEW

File: A2017/tensorflow.ipynb [hunk 1]
```diff
@@ -0,0 +1,892 @@
...
+    "```bash\n",
+    "source activate TODO: WRITEME\n",
+    "```\n",
+    "\n",
+    "To return back to normal, simply use the bash command\n",
+    "\n",
+    "```bash\n",
+    "source deactivate\n",
+    "```\n",
+    "\n",
```
Comment (notoraptor) [inline A2017/tensorflow.ipynb:27]: I think it miss a section to explain how to install tensorflow (e.g. for users who want to use it in their own computers). I suggest to just describe the pip command:
```
pip install tensorflow-gpu
```
https://www.tensorflow.org/install/install_windows

File: A2017/tensorflow.ipynb [hunk 2]
```diff
@@ -0,0 +1,892 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Paradigm\n",
+    "\n",
+    "TensorFlow separates the _definition_ of computation from its _execution_.\n",
+    "\n",
+    "Computation is defined via a [_dataflow_](https://en.wikipedia.org/wiki/Dataflow_programming) _graph_, i.e., a graph where nodes represent units of computation and the edges represent the data consumed or produced by the computation.\n",
```
Comment (notoraptor) [inline A2017/tensorflow.ipynb:63]: About this, I think tensorflow graph is more like Theano graph, ie. a graph with 2 types of nodes: operation nodes and input/output nodes. Tensors are the input/output nodes. Edges just connect an input/output node to an operation node.

File: A2017/tensorflow.ipynb [hunk 3]
```diff
@@ -0,0 +1,892 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 20,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "with tf.Graph().as_default():\n",
```
Comment (notoraptor) [inline A2017/tensorflow.ipynb]: You have not explained `tf.Graph()` before.

I suggest either you explain why you use it, or you write this exercise as a single separate script (with no needs to separate graphs).

File: A2017/tensorflow.ipynb [hunk 4]
```diff
@@ -0,0 +1,892 @@
...
+    "with tf.Session() as session:\n",
+    "    print('The value for r is {}'.format(session.run(r)))\n",
+    "    print('The value for r is {}'.format(session.run(r)))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Combining tensors\n",
```
Comment (notoraptor) [inline A2017/tensorflow.ipynb:473]: In this section, maybe should you mention that some ops have overloaded operators (e.g. `+` for tf.add()).
```python
a = tf.constant(1)
b = tf.constant(2)
c = a + b
```

**Review by notoraptor — CHANGES_REQUESTED** [review]
Hy @vdumoulin , I made a review to suggest quick fixes to your tensorflow tutorial.

Also, some other remarks:
- You still have some TODO/WRTEME in the tutorial.
- There is no section about how to explicitely select a GPU device or a CPU device. Maybe this page could help: https://www.tensorflow.org/tutorials/using_gpu
- There is no section about how to make a symbolic simple loop (like theano scan). But maybe do you consider it as an op that users should discover by browsing the API (`tf.while_loop()`) ?
- There is no section about profiling. Personally, I learned it from this tutorial: https://towardsdatascience.com/howto-profile-tensorflow-1a49fb18073d

I don't know if you will have time to add these sections, or time to present it, but maybe at least the section about GPUs will be required ? @lamblin @nouiz 

Thanks !

**Comment by vdumoulin** [comment]
@notoraptor Thanks a lot for the review! I apologize for being a little last-minute on the tutorial, but I do intend to make it as complete as possible today, so I will definitely consider your suggestions.

File: A2017/tensorflow.ipynb (hunk 1 above)
Comment (vdumoulin) [reply A2017/tensorflow.ipynb:27]: Done via 5c29d47.

File: A2017/tensorflow.ipynb (hunk 2 above)
Comment (vdumoulin) [reply A2017/tensorflow.ipynb:63]: This is the way the [TensorFlow documentation](https://www.tensorflow.org/programmers_guide/graphs) describes it.

Comment (notoraptor) [reply A2017/tensorflow.ipynb:63]: Ok, I see!

File: A2017/tensorflow.ipynb (hunk 4 above)
Comment (vdumoulin) [reply A2017/tensorflow.ipynb:473]: Done via 5a22e52.

File: A2017/tensorflow.ipynb (hunk 3 above)
Comment (vdumoulin) [reply A2017/tensorflow.ipynb]: Good point! Done via 0714eaa.

File: A2017/tensorflow.ipynb [hunk 5]
```diff
@@ -0,0 +1,1357 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In addition to `tf.assign_add`, the `tf.assign_sub` and `tf.assign` functions return ops which decrement and assign a variable's value, respectively.\n",
+    "\n",
+    "## Placeholder\n",
+    "\n",
+    "Oftentimes the computation we define depends on data which we don't yet have. For instance, the output of a neural network depends on a user-defined input which is only specified at runtime.\n",
```
Comment (lamblin) [inline A2017/tensorflow.ipynb:312]: Maybe mention (even orally) that it is for when those values are computed outside of tensorflow

File: A2017/tensorflow.ipynb [hunk 6]
```diff
@@ -0,0 +1,1357 @@
...
+    "    print('The value for r is {}'.format(session.run(r)))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Combining tensors\n",
+    "\n",
+    "Tensors can be combined in various ways using what TensorFlow calls operations, or _ops_. Ops can take zero or more tensors as input and produce zero or more tensors as output, with our without side effects.\n",
```
Comment (lamblin) [inline A2017/tensorflow.ipynb]: "with or without"

File: A2017/tensorflow.ipynb [hunk 7]
```diff
@@ -0,0 +1,1357 @@
...
+    "            # Retrieve the values for x, y, and z, and print them.\n",
+    "            # x_val, y_val, z_val = WRITEME.\n",
+    "            # print('x = {:4.2f}, y = {:4.2f}, z = {:4.2f}'.format(x_val, y_val, z_val))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Solution"
```
Comment (lamblin) [inline A2017/tensorflow.ipynb:748]: If you have time to be fancy, you could put the solution in a different file to hide it, and put the magic command `%load solution/thatfile.py` in a cell instead.

File: A2017/tensorflow.ipynb [hunk 8]
```diff
@@ -0,0 +1,1357 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Control flow\n",
+    "\n",
+    "TensorFlow implements symbolic operations for executing common control flow structures, like if-statements and while loops.\n",
+    "\n",
+    "This is usually where people get confused: despite appearances, there is a *big* difference between a regular control flow statement and its symbolic counterpart. **When working with control flow ops, you should always keep in mind that the code you write _defines the computation graph_, it does not _execute_ that computation.**\n",
```
Comment (lamblin) [inline A2017/tensorflow.ipynb:875]: A way I have heard that described is that it is _metaprogramming_: we are using one language (Python) to build a program in a different language (TF's dataflow). It's not unlike the C preprocessor macros in that sense: there is a big difference between `#if ...` and `if(...)`.

**Review by lamblin — COMMENTED** [review]
Looks good to me up to now, thanks!

File: A2017/tensorflow.ipynb [hunk 9]
```diff
@@ -0,0 +1,1357 @@
...
+    "## Protip: browsing through the TensorFlow API\n",
+    "\n",
+    "The [devdocs.io](http://devdocs.io/) website is an amazing resource to browse through the TensorFlow Python API (as well as many other APIs such as numpy or the Python API itself).\n",
+    "\n",
+    "# Using TensorFlow at MILA\n",
+    "\n",
+    "The most straightforward way to access TensorFlow using the MILA software stack is through the `TODO: WRITEME` conda environment. To activate the `TODO: WRITEME` environment, use the command\n",
+    "\n",
+    "```bash\n",
+    "source activate TODO: WRITEME\n",
```
Comment (notoraptor) [inline A2017/tensorflow.ipynb]: Hi @vdumoulin you still have `todo: writeme` here and above (line 16). Is it expected ?

File: A2017/tensorflow.ipynb [hunk 10]
```diff
@@ -0,0 +1,1357 @@
...
+    "You can find more information on manual device placement in the [TensorFlow documentation](https://www.tensorflow.org/tutorials/using_gpu#manual_device_placement)."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Graph visualization with TensorBoard\n",
+    "\n",
+    "**TODO: WRITEME**"
```
Comment (notoraptor) [inline A2017/tensorflow.ipynb]: Here too !

**Comment by lamblin** [comment]
Also, I know it is last-minute, but maybe it would be worth having an "advanced topics" section mentioning (and linking to) in particular:
- Data management, https://www.tensorflow.org/programmers_guide/datasets
- Estimators, https://www.tensorflow.org/programmers_guide/estimators (esp. for those who could get access to TPUs)
- Eager mode, https://github.com/tensorflow/tensorflow/tree/master/tensorflow/contrib/eager

**Comment by vdumoulin** [comment]
@lamblin it's been a while since I last kept up-to-date with the latest TensorFlow features, and I certainly won't be able to discuss them in the tutorial, but sure, I can link to them!

File: A2017/tensorflow.ipynb (hunk 5 above)
Comment (vdumoulin) [reply A2017/tensorflow.ipynb:312]: Will do.

File: A2017/tensorflow.ipynb (hunk 6 above)
Comment (vdumoulin) [reply A2017/tensorflow.ipynb]: Done via 0ae5653.

File: A2017/tensorflow.ipynb (hunk 7 above)
Comment (vdumoulin) [reply A2017/tensorflow.ipynb:748]: Great idea! Done via ce91418.

File: A2017/tensorflow.ipynb (hunk 8 above)
Comment (vdumoulin) [reply A2017/tensorflow.ipynb:875]: That's a good analogy as well!

File: A2017/tensorflow.ipynb (hunk 10 above)
Comment (vdumoulin) [reply A2017/tensorflow.ipynb]: Done via 9bec371.

**Comment by vdumoulin** [comment]
@lamblin @notoraptor I think I covered all the points you raised in your reviews, would you mind having another look?

**Comment by notoraptor** [comment]
It looks good for me !

**Comment by lamblin** [comment]
Merged, thanks a lot!

**Pull request merged**